libraries/AC_AutoTune/AC_AutoTune.cpp

```diff
@@ -1710,56 +1241,585 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
+                break;
+            }
+        }
+    } else {
+        // Testing rate P and D gains so will set body-frame rate targets.
+        // Rate controller will use existing body-frame rates and convert to motor outputs
+        // for all axes except the one we override here.
+        switch (test_axis) {
+        case ROLL:
+            // override body-frame roll rate
+            attitude_control->rate_bf_roll_target(dir_sign * target_rate + start_rate);
+            break;
+        case PITCH:
+            // override body-frame pitch rate
+            attitude_control->rate_bf_pitch_target(dir_sign * target_rate + start_rate);
+            break;
+        case YAW:
+            // override body-frame yaw rate
+            attitude_control->rate_bf_yaw_target(dir_sign * target_rate + start_rate);
+            break;
+        }
+    }
+
+    // capture this iterations rotation rate and lean angle
+    float gyro_reading = 0;
+    switch (test_axis) {
+    case ROLL:
+        gyro_reading = ahrs_view->get_gyro().x;
+        lean_angle = dir_sign * (ahrs_view->roll_sensor - (int32_t)start_angle);
+        break;
+    case PITCH:
+        gyro_reading = ahrs_view->get_gyro().y;
+        lean_angle = dir_sign * (ahrs_view->pitch_sensor - (int32_t)start_angle);
+        break;
+    case YAW:
+        gyro_reading = ahrs_view->get_gyro().z;
+        lean_angle = dir_sign * wrap_180_cd(ahrs_view->yaw_sensor-(int32_t)start_angle);
+        break;
+    }
+
+    // Add filter to measurements
+    float filter_value;
+    switch (tune_type) {
+    case SP_DOWN:
+    case SP_UP:
+        filter_value = dir_sign * (ToDeg(gyro_reading) * 100.0f);
+        break;
+    default:
+        filter_value = dir_sign * (ToDeg(gyro_reading) * 100.0f - start_rate);
+        break;
+    }
+    rotation_rate = rotation_rate_filt.apply(filter_value,
+                    AP::scheduler().get_loop_period_s());
+
+    switch (tune_type) {
+    case RD_UP:
+    case RD_DOWN:
+        twitching_test_rate(rotation_rate, target_rate, test_rate_min, test_rate_max);
+        twitching_measure_acceleration(test_accel_max, rotation_rate, rate_max);
+        twitching_abort_rate(lean_angle, rotation_rate, abort_angle, test_rate_min);
+        break;
+    case RP_UP:
+        twitching_test_rate(rotation_rate, target_rate*(1+0.5f*aggressiveness), test_rate_min, test_rate_max);
+        twitching_measure_acceleration(test_accel_max, rotation_rate, rate_max);
+        twitching_abort_rate(lean_angle, rotation_rate, abort_angle, test_rate_min);
+        break;
+    case SP_DOWN:
+    case SP_UP:
+        twitching_test_angle(lean_angle, rotation_rate, target_angle*(1+0.5f*aggressiveness), test_angle_min, test_angle_max, test_rate_min, test_rate_max);
+        twitching_measure_acceleration(test_accel_max, rotation_rate - dir_sign * start_rate, rate_max);
+        break;
+    default:
+        break;
+    }
+}
+
+void AC_AutoTune::rate_ff_test_init()
```

Comment:
rate_ff_test_init is removed from this PR so it is not a blocker but I think this should eventually go into the AC_AutoTune_Heli class because it's not planning to be used by multicopters (as far as I know).

Reply:
As discussed, we'll leave these in the parent class until I finish the heli autotune and clean up things then